pkg/sql/rollback.go

```diff
@@ -0,0 +1,226 @@
+		}
+
+		if !tables[i].IsPhysicalTable() {
+			return errors.Errorf("cannot revert virtual table %s", tables[i].GetName())
+		}
+		spans = append(spans, tables[i].TableSpan(execCfg.Codec))
+	}
+
+	for i := range tables {
+		// This is a) rare and b) probably relevant if we are looking at logs so it
+		// probably makes sense to log it without a verbosity filter.
+		log.Infof(ctx, "reverting table %s (%d) to time %v", tables[i].GetName(), tables[i].GetID(), targetTime)
+	}
+
+	// TODO(dt): pre-split requests up using a rangedesc cache and run batches in
+	// parallel (since we're passing a key limit, distsender won't do its usual
+	// splitting/parallel sending to separate ranges).
+	for len(spans) != 0 {
+		var b kv.Batch
+		for _, span := range spans {
+			b.AddRawRequest(&roachpb.RevertRangeRequest{
+				RequestHeader: roachpb.RequestHeader{
+					Key:    span.Key,
+					EndKey: span.EndKey,
+				},
+				TargetTime:                          targetTime,
+				IgnoreGcThreshold:                   ignoreGCThreshold,
+				EnableTimeBoundIteratorOptimization: true, // NB: Must set for 22.1 compatibility.
+			})
+		}
+		b.Header.MaxSpanRequestKeys = batchSize
+
+		if err := db.Run(ctx, &b); err != nil {
+			return err
+		}
+
+		spans = spans[:0]
+		for _, raw := range b.RawResponse().Responses {
+			r := raw.GetRevertRange()
+			if r.ResumeSpan != nil {
+				if !r.ResumeSpan.Valid() {
+					return errors.Errorf("invalid resume span: %s", r.ResumeSpan)
+				}
+				spans = append(spans, *r.ResumeSpan)
+			}
+		}
+	}
+
+	return nil
+}
+
+// DeleteTableWithPredicate issues a series of point and range tombstones over a subset of keys in a
+// table that match the passed-in predicate.
+//
+// This function will error, without a resume span, if it encounters an intent
+// in the span. The caller should resolve these intents by retrying the
+// function. To prevent errors, the caller should only pass a table that will
+// not see new writes during this bulk delete operation (e.g. on a span that's
+// part of an import rollback).
+//
+// NOTE: this function will issue tombstones on keys with versions that match
+// the predicate, in contrast to RevertRange, which rolls back a table to a
+// certain timestamp. For example, if a key gets a single update after the
+// predicate.startTime, DeleteTableWithPredicate would delete that key, while
+// RevertRange would revert that key to its state before the update.
+func DeleteTableWithPredicate(
+	ctx context.Context,
+	db *kv.DB,
+	codec keys.SQLCodec,
+	distSender *kvcoord.DistSender,
+	table catalog.TableDescriptor,
+	predicates roachpb.DeleteRangePredicates,
+	batchSize int64,
+) error {
+
+	log.Infof(ctx, "deleting data for table %d with predicate %s", table.GetID(), predicates.String())
+	tableKey := roachpb.RKey(codec.TablePrefix(uint32(table.GetID())))
+	tableSpan := roachpb.RSpan{Key: tableKey, EndKey: tableKey.PrefixEnd()}
+
+	// TODO (msbutler): tune these
+	rangesPerBatch := int64(100)
```

Comment:
Is 100 a bit much? Maybe 10 or 20? Do we have any experience with similar settings elsewhere?

Reply:
@dt any prior here? If not, i can ask KV, or tune this during stability.